Improved WebGL architecture

MortimerGoro opened the pull request:
<!-- Please describe your changes on the following line: -->

Info about the big picture and the goals of the WebGL refactor in this thread: https://groups.google.com/forum/#!topic/mozilla.dev.servo/0WMGz60kKzQ

I tried to reduce this PR as much as possible as requested in the thread. I'll do separate PRs for other features (e.g.: Batch messages or use shared memory to improve frame times) or fixes.

Some tips to ease the review process:  
- Most changes in DOM objects follow the same pattern (remove CanvasMsg wrapper and use the new sender method). 
- WebGLCommands are the same ones as before (moved from webrender_api). So those lines are already reviewed.
- See WebGL traits in [components/canvas_traits/webgl.rs](https://github.com/servo/servo/pull/17891/files#diff-8701045d01505418701d0631d4d45562)
- See WebGLThread and WR External Image bridge in [components/canvas/webgl_thread.rs](https://github.com/servo/servo/pull/17891/files#diff-281554879f39a2a041f7a69d442a5d2e)
- The implementation submitted in this PR creates a single `WebGLThread` for all ScriptThread/Pipelines. See that in [components/canvas/webgl_mode/inprocess.rs](https://github.com/servo/servo/pull/17891/files#diff-250070c6c5a38c7f9fa0f5b3c101f68b)

The conformance tests will help to guarantee that we don't miss anything.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17891)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/constellation/pipeline.rs, components/constellation/lib.rs, components/constellation/Cargo.toml, components/constellation/constellation.rs
 * @kichjang: components/script/dom/webglframebuffer.rs, components/script/dom/webglshader.rs, components/script/dom/webgl_validations/tex_image_2d.rs, components/script/dom/htmlcanvaselement.rs, components/script/dom/window.rs and 24 more
 * @fitzgen: components/script/dom/webglframebuffer.rs, components/script/dom/webglshader.rs, components/script/dom/webgl_validations/tex_image_2d.rs, components/script/dom/htmlcanvaselement.rs, components/script/dom/window.rs and 24 more
 * @emilio: components/layout/display_list_builder.rs, components/script/dom/webglframebuffer.rs, components/script/dom/webglshader.rs, components/script/dom/webgl_validations/tex_image_2d.rs, components/layout/webrender_helpers.rs and 14 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify gfx, layout, and script code, but no tests are modified. Please consider adding a test!

stshine commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 1cee8781fe62d902d23b6700c90c229bd1419695 with merge dc3313819afcc36fa1c6df5a41ebaa795fa17327...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5228)

jdm commented:
```
  ▶ FAIL [expected PASS] /_mozilla/mozilla/webgl/tex_image_2d_mipmap.html
  └   → /_mozilla/mozilla/webgl/tex_image_2d_mipmap.html 1017abb40a2237d2ac16e329c9754cbedcf30e1c
/_mozilla/mozilla/webgl/tex_image_2d_simple_ref.html 06711e6d8ba9520a7c0476d41130022d3e609160
Testing 1017abb40a2237d2ac16e329c9754cbedcf30e1c == 06711e6d8ba9520a7c0476d41130022d3e609160
```

Also, all of our multiprocess tests panicked with `internal error: entered unreachable code (thread Constellation, at /home/servo/buildbot/slave/linux-rel-wpt/build/components/canvas_traits/webgl_channel/mpsc.rs:76)`. We run them like so: `test-wpt --release --binary-arg=--multiprocess --processes 24 eventsource`

asajeffrey commented:
Yay, our multiprocess tests worked!

MortimerGoro commented:
> ▶ FAIL [expected PASS] /_mozilla/mozilla/webgl/tex_image_2d_mipmap.html
>   └   → /_mozilla/mozilla/webgl/tex_image_2d_mipmap.html 

Fixed. The `webrender_api::ImageDescriptor.alpha` property was not correctly set, leading to a black background instead of a transparent one when the webgl canvas is composited. To fix this I used the WebGLContextAttributes::alpha attribute to correctly set the is_opaque `webrender_api::ImageDescriptor.alpha` property.

`Also, all of our multiprocess tests panicked with internal error: entered unreachable code (thread Constellation, at /home/servo/buildbot/slave/linux-rel-wpt/build/components/canvas_traits/webgl_channel/mpsc.rs:76). We run them like so: ./mach test-wpt --release --binary-arg=--multiprocess --processes 24 eventsource`

Thanks for the info! I fixed the unreachable code crash and all multiprocess tests pass now.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17892) made this pull request unmergeable. Please resolve the merge conflicts.

MortimerGoro commented:
Rebased. The latest WR update includes https://github.com/servo/webrender/pull/1524 which was required to fix a failing webgl test.

All tests should pass now.

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 931826b7f9e5e321ab3850efa2ede8cb0e2be1a9 with merge 1f2d551d81e21bad808aedeadddf5aa3e48dd100...

bors-servo commented:
:snowman: The build was interrupted to prioritize another pull request.

bors-servo commented:
:hourglass: Trying commit 931826b7f9e5e321ab3850efa2ede8cb0e2be1a9 with merge e0645d0378b7899cccadf7123b4015d4be1b73e0...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/8118), [arm32](http://build.servo.org/builders/arm32/builds/8341), [arm64](http://build.servo.org/builders/arm64/builds/8374), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8460), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5222), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5256), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8455), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/932), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/931), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5157), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5215), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/934), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/938), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3699)
State: approved= try=True

MortimerGoro commented:
Rebased to fix the conflict. Why does it still show `needs-rebase` label?

jdm commented:
The github API hooks have never provided real mergeability information to the bot, so we have never been able to automatically remove the needs-rebase label.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17923) made this pull request unmergeable. Please resolve the merge conflicts.

larsbergstrom commented:
@jdm Are you the right person to review this, or should we have somebody else look at it? Thanks!

emilio commented:
I plan to look at this soon fwiw, just have to find the time .

Obviously help from other people to review such large patches is way
welcome.

On 08/05/2017 05:37 PM, Lars Bergstrom wrote:
> @jdm <https://github.com/jdm> Are you the right person to review this,
> or should we have somebody else look at it? Thanks!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/servo/servo/pull/17891#issuecomment-320450385>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQwusRdP_ZWpjdM7eIkpWhT2RcW28Ukks5sVIw-gaJpZM4Oliuj>.
>

nox commented:
@MortimerGoro Any chance you could split that PR into multiple self-contained commits?

MortimerGoro commented:
@nox Yes, I can't split into separate PRs but I can rewrite the PR-history with self-contained commits.

I'll update the PR later. We can squash all the commits again after it's reviewed.

jdm commented:
I've got a lot of things to review and I'm on the hook for some stylo performance investigations. If @glennw or @pcwalton could review this instead, that would be a big help.

MortimerGoro commented:
I split the PR into 23 self-contained commits ;)

glennw commented:
I will look through it today and review the parts I can.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #18002) made this pull request unmergeable. Please resolve the merge conflicts.

glennw commented:
Reviewed 2 of 2 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 2 of 2 files at r6, 9 of 9 files at r7, 3 of 3 files at r8, 4 of 4 files at r9, 6 of 6 files at r10, 2 of 2 files at r11, 1 of 1 files at r12, 1 of 1 files at r13, 7 of 7 files at r14, 1 of 1 files at r15, 1 of 1 files at r16, 1 of 1 files at r17, 2 of 2 files at r18, 1 of 1 files at r19, 3 of 3 files at r20, 9 of 9 files at r21, 1 of 1 files at r22.
Review status: 53 of 54 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[components/canvas/Cargo.toml, line 23 at r20](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3Wvm5-Vq9O0J3BDmv:-Kr3Wvm6Fn6ZoFXjkAmt:b-tlre3p) ([raw file](https://github.com/servo/servo/blob/af55f03e1b66508be7b7aa3104c20c040a4daeb2/components/canvas/Cargo.toml#L23)):*
> ```TOML
> num-traits = "0.1.32"
> offscreen_gl_context = { version = "0.11", features = ["serde"] }
> webrender = {git = "https://github.com/servo/webrender"}
> ```

Ouch - I think this will mean that modifying WR will mean a full rebuild of servo each time. Since the dep chain would now be WR -> canvas -> script -> servo - is that right?

I guess it's related to the external image APIs? If so, perhaps we can move the required type to the webrender_api crate? (which could be done as a follow up, so long as we have a clear plan).

---

*[components/canvas/webgl_thread.rs, line 30 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3SnzMM5gkVFR9-ZqZ:-Kr3SnzNtxLAVjj2TOnN:bxyx2zn) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L30)):*
> ```Rust
>     webrender_api: webrender_api::RenderApi,
>     /// Map of live WebGLContexts.
>     contexts: HashMap<WebGLContextId, GLContextWrapper>,
> ```

We should probably use Fx (or Fnv) as the hasher here. Can be done as a follow up though.

---

*[components/canvas/webgl_thread.rs, line 32 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3SyM3-d4FZKOc0QqC:-Kr3SyM3-d4FZKOc0QqD:buo9nmt) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L32)):*
> ```Rust
>     contexts: HashMap<WebGLContextId, GLContextWrapper>,
>     /// Cached information for WebGLContexts.
>     cached_context_info: HashMap<WebGLContextId, WebGLContextInfo>,
> ```

Same here

---

*[components/canvas/webgl_thread.rs, line 55 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3T3Olo8OHyDCirlMC:-Kr3T3Olo8OHyDCirlMD:bwka87k) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L55)):*
> ```Rust
>             current_bound_webgl_context_id: None,
>             next_webgl_id: 0,
>             webvr_compositor: webvr_compositor,
> ```

nit: can use new init syntax here to avoid duplication.

---

*[components/canvas/webgl_thread.rs, line 159 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3TPINLyRwO1n-h-z0:-Kr3TPIOMGW4ICy7Q3MX:baspv4s) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L159)):*
> ```Rust
>     fn handle_lock(&mut self, context_id: WebGLContextId, sender: WebGLSender<(u32, Size2D<i32>)>) {
>         let ctx = &self.contexts[&context_id];
>         if Some(context_id) != self.current_bound_webgl_context_id {
> ```

Perhaps add a method that does the lazy binding of the context to avoid duplication.

---

*[components/canvas/webgl_thread.rs, line 170 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3TajGVzZ0dxxvENEW:-Kr3TajGVzZ0dxxvENEX:b-filowx) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L170)):*
> ```Rust
>     }
> 
>     /// Handles a unlock external callback received from webrender::ExternalImageHandler
> ```

uber nit: a -> an

---

*[components/canvas/webgl_thread.rs, line 181 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3TfJ1-NcyyrRFnF5x:-Kr3TfJ2pCQ3Q4rKcJgb:bmflkdg) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L181)):*
> ```Rust
>             // glFlush must be called before glWaitSync.
>             ctx.gl().flush();
>             // Waint until the GLSync object is signaled.
> ```

waint -> wait

---

*[components/canvas/webgl_thread.rs, line 212 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3Tl9XRiCIy_nwaLkj:-Kr3Tl9XRiCIy_nwaLkk:byjfxlp) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L212)):*
> ```Rust
>                 self.next_webgl_id += 1;
>                 self.contexts.insert(id, ctx);
>                 self.cached_context_info.insert(id, WebGLContextInfo {
> ```

Could use shorter init syntax here.

---

*[components/canvas/webgl_thread.rs, line 252 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3TsTJ0rLvFaDSo75_:-Kr3TsTJ0rLvFaDSo75a:b-z44uif) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L252)):*
> ```Rust
>                 info.texture_id = texture_id;
>                 info.size = real_size;
>                 // WR doesn't support resizing and requires to create a new `ImageKey`.
> ```

WR does now support resizing with the same image key (as of a week or so ago). We could fix this as a follow up though if you prefer.

---

*[components/canvas/webgl_thread.rs, line 386 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3ULWXZUGEOJfrbeB9:-Kr3ULWXZUGEOJfrbeBA:b-lm62gn) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L386)):*
> ```Rust
>     }
> 
>     /// Creates a `webrender_api::ImageKey` that uses raw pixels.
> ```

Creates -> updates

---

*[components/canvas/webgl_thread.rs, line 407 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3UTP6hia8AcKUm571:-Kr3UTP6hia8AcKUm572:bk1ufjm) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L407)):*
> ```Rust
>             height: size.height as u32,
>             stride: None,
>             format: if alpha { webrender_api::ImageFormat::RGB8 } else { webrender_api::ImageFormat::BGRA8 },
> ```

This format seems around the wrong way?

---

*[components/canvas/webgl_thread.rs, line 764 at r17](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3UsBxTRuxz_hLBYG1:-Kr3UsBxTRuxz_hLBYG2:bef1het) ([raw file](https://github.com/servo/servo/blob/24ddebdfbf9634b92027fe26622e172a7e2f0e9c/components/canvas/webgl_thread.rs#L764)):*
> ```Rust
>         }
> 
>         // TODO: update test expectations in order to enable debug assertions
> ```

Does this need to be resolved before merging?

---

*[components/canvas/webgl_mode/inprocess.rs, line 40 at r18](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3VFKMQej5mSLi9nnK:-Kr3VFKMQej5mSLi9nnL:b-4v3jxd) ([raw file](https://github.com/servo/servo/blob/62dcfe47f832877b63006df168a20ccafb2331d9/components/canvas/webgl_mode/inprocess.rs#L40)):*
> ```Rust
>     }
> 
>     /// Sends a exit message to close all the existing `WebGLThread`s.
> ```

Comment needs updating.

---

*[components/canvas_traits/webgl.rs, line 15 at r2](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3LZ-sj9ZQmb59kXEv:-Kr3LZ-t9w9cjsGgdHII:bbokzdx) ([raw file](https://github.com/servo/servo/blob/3707aa557163a25eab9a97a38e1ddd7d23d1c224/components/canvas_traits/webgl.rs#L15)):*
> ```Rust
> /// Receiver type used in WebGLCommands.
> pub use ::webgl_channel::WebGLReceiver;
> /// Result type for rend()/recv() calls in in WebGLCommands.
> ```

rend -> send

---

*[components/canvas_traits/webgl.rs, line 37 at r2](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3LlpcZe-fnLTCuoN-:-Kr3LlpcZe-fnLTCuoN0:b-l4un8t) ([raw file](https://github.com/servo/servo/blob/3707aa557163a25eab9a97a38e1ddd7d23d1c224/components/canvas_traits/webgl.rs#L37)):*
> ```Rust
>     /// Runs a WebVRCommand in a specific WebGLContext.
>     WebVRCommand(WebGLContextId, WebVRCommand),
>     /// Locks a specific WebGLContext.
> ```

Perhaps expand a bit more on what lock means and what the semantics are?

---

*[components/canvas_traits/webgl.rs, line 43 at r2](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3LrJE2lJQeLVLg_HM:-Kr3LrJE2lJQeLVLg_HN:b6zzob8) ([raw file](https://github.com/servo/servo/blob/3707aa557163a25eab9a97a38e1ddd7d23d1c224/components/canvas_traits/webgl.rs#L43)):*
> ```Rust
>     /// Creates or updates the image keys required for WebRender.
>     UpdateWebRenderImage(WebGLContextId, WebGLSender<webrender_api::ImageKey>),
>     /// Fress al resources and closes the thread.
> ```

nit: spelling

---

*[components/layout/display_list_builder.rs, line 1997 at r14](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3QslRnC7nTTZQKWaC:-Kr3QslSEkHrOeetnoSC:b-r7vhgx) ([raw file](https://github.com/servo/servo/blob/e4913a114a86cd1c1afbcc79b6650fb620e10e34/components/layout/display_list_builder.rs#L1997)):*
> ```Rust
>                                 ipc_renderer.send(CanvasMsg::FromLayout(
>                                     FromLayoutMsg::SendData(sender))).unwrap();
>                                 (receiver.recv().unwrap().image_key, PixelFormat::BGRA8)
> ```

Is this likely to cause stalls? It doesn't seem ideal to have a sync msg here.

---

*[components/script/dom/canvasgradient.rs, line 5 at r10](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3PLCwJLH9haOmfuRQ:-Kr3PLCwJLH9haOmfuRR:bisxprl) ([raw file](https://github.com/servo/servo/blob/c15d7dbc9bd6fd1586b1f90c4c7d19ea5a3fd86a/components/script/dom/canvasgradient.rs#L5)):*
> ```Rust
>  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
> 
> use canvas_traits::canvas::{CanvasGradientStop, FillOrStrokeStyle, LinearGradientStyle, RadialGradientStyle};
> ```

I guess these could be exposed as pub type in the lib.rs still, but it's fine as is.

---

*[components/script/dom/htmlcanvaselement.rs, line 229 at r11](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3PylKHLBBm6pM9iUS:-Kr3PylKHLBBm6pM9iUT:bpu4jly) ([raw file](https://github.com/servo/servo/blob/7a782d0e979a619d9ed96b0ae8ad99dc633cd941/components/script/dom/htmlcanvaselement.rs#L229)):*
> ```Rust
>             },
>             Some(&CanvasContext::WebGL(_)) => {
>                 // TODO: add a method in WebGLRenderingContext to get the pixels.
> ```

Do we need a GH issue to track this?

---

*[components/script/dom/htmlcanvaselement.rs, line 241 at r11](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3PwCGBcuRkROXPMQr:-Kr3PwCHQRbWZwEHeo4A:b-5h52wl) ([raw file](https://github.com/servo/servo/blob/7a782d0e979a619d9ed96b0ae8ad99dc633cd941/components/script/dom/htmlcanvaselement.rs#L241)):*
> ```Rust
> 
>     pub fn is_webgl(&self) -> bool {
>         if let Some(ref context) = *self.context.borrow() {
> ```

Perhaps tidier as a match statement?

---

*[components/script/dom/webglrenderingcontext.rs, line 721 at r7](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3OE027we2ZIPpsYe8:-Kr3OE03_G75yQfi8_yk:bsld7m6) ([raw file](https://github.com/servo/servo/blob/e302448bcfda06285b2cf3765bad2bf23cc07d52/components/script/dom/webglrenderingcontext.rs#L721)):*
> ```Rust
>             ImageDataOrHTMLImageElementOrHTMLCanvasElementOrHTMLVideoElement::HTMLCanvasElement(canvas) => {
>                 if let Some((mut data, size)) = canvas.fetch_all_data() {
>                     if !canvas.is_webgl() {
> ```

It's not clear to me why this is needed now - but might just be because I haven't got far enough into the review yet.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/17891)*
<!-- Sent from Reviewable.io -->

glennw commented:
@MortimerGoro Wow, this looks great!

Review comments above - most of them are very minor. I think there's a couple of logic errors, and we should figure out a plan for the WR dependency that canvas now includes.

It'd be good for someone else to look over it if they have time too...

MortimerGoro commented:
Review status: 53 of 54 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[components/canvas/Cargo.toml, line 23 at r20](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3Wvm5-Vq9O0J3BDmv:-Kr5LRoT9cNnqkchhJzb:b8fue1f) ([raw file](https://github.com/servo/servo/blob/af55f03e1b66508be7b7aa3104c20c040a4daeb2/components/canvas/Cargo.toml#L23)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Ouch - I think this will mean that modifying WR will mean a full rebuild of servo each time. Since the dep chain would now be WR -> canvas -> script -> servo - is that right?

I guess it's related to the external image APIs? If so, perhaps we can move the required type to the webrender_api crate? (which could be done as a follow up, so long as we have a clear plan).
</blockquote></details>

Right now canvas component is quite fast to compile, it only triggers a constellation compilation.  Script depends on canvas_traits (not canvas). 

 I like the idea of moving the external_imagi_api to webrender_api. We can avoid WR dependency here (will as a follow up)

---

*[components/canvas/webgl_thread.rs, line 30 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3SnzMM5gkVFR9-ZqZ:-Kr5POC-R_3T-jMlVA7d:bxge6k7) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L30)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

We should probably use Fx (or Fnv) as the hasher here. Can be done as a follow up though.
</blockquote></details>

Ok, will fix this as a follow up.

---

*[components/canvas/webgl_thread.rs, line 32 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3SyM3-d4FZKOc0QqC:-Kr5POt0lkaTndsrv6E3:b7ybaur) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L32)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Same here
</blockquote></details>

Ok, will fix this as a follow up.

---

*[components/canvas/webgl_thread.rs, line 55 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3T3Olo8OHyDCirlMC:-Kr5sM9ZFM3TiCeT52re:b-896fix) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L55)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

nit: can use new init syntax here to avoid duplication.
</blockquote></details>

Done.

---

*[components/canvas/webgl_thread.rs, line 159 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3TPINLyRwO1n-h-z0:-Kr6KrzTU0yVgeFdVyB2:b-9idqju) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L159)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Perhaps add a method that does the lazy binding of the context to avoid duplication.
</blockquote></details>

Ok, I'll add the method. Maybe I'll do as a follow up because rust is giving self mutability borrow errors (it doesn't detect partial borrowing of different fields).

---

*[components/canvas/webgl_thread.rs, line 170 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3TajGVzZ0dxxvENEW:-Kr5sVvyy_MGp1gJkj3b:b-896fix) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L170)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

uber nit: a -> an
</blockquote></details>

Done.

---

*[components/canvas/webgl_thread.rs, line 181 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3TfJ1-NcyyrRFnF5x:-Kr5sbpBOb9Uh21XKv44:b-896fix) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L181)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

waint -> wait
</blockquote></details>

Done.

---

*[components/canvas/webgl_thread.rs, line 212 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3Tl9XRiCIy_nwaLkj:-Kr5sqW3OMhbh-7iXvce:b-896fix) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L212)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Could use shorter init syntax here.
</blockquote></details>

Done.

---

*[components/canvas/webgl_thread.rs, line 252 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3TsTJ0rLvFaDSo75_:-Kr5NxPVAPhOQZKAT81s:b7ybaur) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L252)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

WR does now support resizing with the same image key (as of a week or so ago). We could fix this as a follow up though if you prefer.
</blockquote></details>

Ok, will fix this as a follow up.

---

*[components/canvas/webgl_thread.rs, line 386 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3ULWXZUGEOJfrbeB9:-Kr5sxhv2LMhVtcljzMf:b-896fix) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L386)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Creates -> updates
</blockquote></details>

Done.

---

*[components/canvas/webgl_thread.rs, line 407 at r16](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3UTP6hia8AcKUm571:-Kr5t8Rhi2aUanSNB2E9:b-896fix) ([raw file](https://github.com/servo/servo/blob/34198c0932a00c27950d5dd6e68c895d3d085f91/components/canvas/webgl_thread.rs#L407)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

This format seems around the wrong way?
</blockquote></details>

Done.

---

*[components/canvas/webgl_thread.rs, line 764 at r17](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3UsBxTRuxz_hLBYG1:-Kr5OJitKLdZvyDJSxD6:bxjvlv) ([raw file](https://github.com/servo/servo/blob/24ddebdfbf9634b92027fe26622e172a7e2f0e9c/components/canvas/webgl_thread.rs#L764)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Does this need to be resolved before merging?
</blockquote></details>

I will do as a follow up. The problem is that in ci,  debug assertions are enabled on linux but not on mac.  I asked the reason of this. It seems that mac builders are slow and debug assertions are disabled for performance. That would require different test expectations for osx and linux.

---

*[components/canvas/webgl_mode/inprocess.rs, line 40 at r18](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3VFKMQej5mSLi9nnK:-Kr6BBfqKfJ7Vy5Fiy8-:b-896fix) ([raw file](https://github.com/servo/servo/blob/62dcfe47f832877b63006df168a20ccafb2331d9/components/canvas/webgl_mode/inprocess.rs#L40)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Comment needs updating.
</blockquote></details>

Done.

---

*[components/canvas_traits/webgl.rs, line 15 at r2](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3LZ-sj9ZQmb59kXEv:-Kr5tZYTBJ63TxXUwpsO:b-896fix) ([raw file](https://github.com/servo/servo/blob/3707aa557163a25eab9a97a38e1ddd7d23d1c224/components/canvas_traits/webgl.rs#L15)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

rend -> send
</blockquote></details>

Done.

---

*[components/canvas_traits/webgl.rs, line 37 at r2](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3LlpcZe-fnLTCuoN-:-Kr6EqjWnkhPGDeD1zuZ:b-896fix) ([raw file](https://github.com/servo/servo/blob/3707aa557163a25eab9a97a38e1ddd7d23d1c224/components/canvas_traits/webgl.rs#L37)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Perhaps expand a bit more on what lock means and what the semantics are?
</blockquote></details>

Done.

---

*[components/canvas_traits/webgl.rs, line 43 at r2](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3LrJE2lJQeLVLg_HM:-Kr5thpXvzD21Erq_vfb:b-896fix) ([raw file](https://github.com/servo/servo/blob/3707aa557163a25eab9a97a38e1ddd7d23d1c224/components/canvas_traits/webgl.rs#L43)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

nit: spelling
</blockquote></details>

Done.

---

*[components/layout/display_list_builder.rs, line 1997 at r14](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3QslRnC7nTTZQKWaC:-Kr5x3dPBtRg2M6NQGSK:bm0zdfg) ([raw file](https://github.com/servo/servo/blob/e4913a114a86cd1c1afbcc79b6650fb620e10e34/components/layout/display_list_builder.rs#L1997)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Is this likely to cause stalls? It doesn't seem ideal to have a sync msg here.
</blockquote></details>

Yes, but it's a problem related to Canvas2DRenderingContext, which will require another big refactor like the one done for WebGL.  I fixed the sync call for WebGL. I'll create a github issue the Canvas2DRenderingContext. 

---

*[components/script/dom/canvasgradient.rs, line 5 at r10](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3PLCwJLH9haOmfuRQ:-Kr5zxLUzrrR8pBE1E3V:b-v79oup) ([raw file](https://github.com/servo/servo/blob/c15d7dbc9bd6fd1586b1f90c4c7d19ea5a3fd86a/components/script/dom/canvasgradient.rs#L5)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

I guess these could be exposed as pub type in the lib.rs still, but it's fine as is.
</blockquote></details>

I split into a different module to separate the types from Canvas2D and WebGL (there are a lot of traits for both)

---

*[components/script/dom/htmlcanvaselement.rs, line 229 at r11](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3PylKHLBBm6pM9iUS:-Kr5qYDXfGVANV7qeHID:b5sevop) ([raw file](https://github.com/servo/servo/blob/7a782d0e979a619d9ed96b0ae8ad99dc633cd941/components/script/dom/htmlcanvaselement.rs#L229)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Do we need a GH issue to track this?
</blockquote></details>

I have the code ready but wanted to split into another PR (because some test expectations need to be updated too). Will add this to the list of follow ups ;)

---

*[components/script/dom/htmlcanvaselement.rs, line 241 at r11](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3PwCGBcuRkROXPMQr:-Kr633UwXc7Jx4gCTjtM:b-7ti6as) ([raw file](https://github.com/servo/servo/blob/7a782d0e979a619d9ed96b0ae8ad99dc633cd941/components/script/dom/htmlcanvaselement.rs#L241)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Perhaps tidier as a match statement?
</blockquote></details>

I decided to remove the is_webgl() method. I will add it together with te code required to get webgl canvas pixels in fetch_all_data() (follow up task)

---

*[components/script/dom/webglrenderingcontext.rs, line 721 at r7](https://reviewable.io:443/reviews/servo/servo/17891#-Kr3OE027we2ZIPpsYe8:-Kr62SaELC5jEfxj3vzy:bclxwc3) ([raw file](https://github.com/servo/servo/blob/e302448bcfda06285b2cf3765bad2bf23cc07d52/components/script/dom/webglrenderingcontext.rs#L721)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

It's not clear to me why this is needed now - but might just be because I haven't got far enough into the review yet.
</blockquote></details>

You are right, it's not required now. I will add it together with te code required to get webgl canvas pixels in fetch_all_data() (follow up task)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/17891)*
<!-- Sent from Reviewable.io -->

MortimerGoro commented:
Thanks for the review @glennw!

I fixed some of the nits. This is a list of follow ups (I will submit them as a separate PRs or create github issues):

- Use Fx (or Fnv) as the hasher
- WR now supports updating images: Update ImageKeys instead of creating new ones.
- Use debug assertions to check WebGL errors
- Create a github issue with the Canvas2DRenderingContext refactor recommendation.
- Add WebGL read pixels in the htmlcanvaselement::fetch_all_data() methods and update test expectations
- Move ExternalImageHandler to webrender_api
- Avoid some minor duplication: Add a method to get & do the lazy binding of the WebGLContext.

emilio commented on components/canvas/gl_context.rs:
```diff
@@ -0,0 +1,203 @@
+}
+
+#[allow(unsafe_code)]
+unsafe impl Send for GLContextFactory {}
```
These deserve a comment. Why do they need to be `Send` + `Sync`? the GL handles are definitely not `Send`...

emilio replied:
If this is just so they can be moved from the main thread to the WebGL thread manager on initialization, what about creating a wrapper that implement those but doesn't expose any API?

MortimerGoro replied:
Send/Sync is not required now. I removed the impls. We might need them in the future if we add more optimized channels to reduce cost per WebGL call even more, but that would go in a different PR with it's explanations and docs.

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -1035,17 +1039,38 @@ impl WebGLRenderingContext {
         receiver.recv().unwrap()
     }
+
+    fn handle_layout(&self) -> webrender_api::ImageKey {
```
nit: `handle_layout` seems like it's handling some kind of command or what not. Instead it's _returning_ a handle. What about renaming it to `layout_handle`?

MortimerGoro replied:
Done

emilio commented on components/canvas/gl_context.rs:
```diff
@@ -0,0 +1,203 @@
+use std::sync::{Arc, Mutex};
+use super::webgl_thread::WebGLImpl;
+
+pub enum GLContextFactory {
```
This needs docs about what's its purpose. In particular, this needs to explain that this is used to create shared GL contexts with the "main thread" context.

MortimerGoro replied:
Done.

emilio commented on components/canvas/webgl_thread.rs:
```diff
@@ -0,0 +1,1191 @@
+
+    /// Handles a generic WebGLMsg message
+    #[inline]
+    pub fn handle_msg(&mut self, msg: WebGLMsg, webgl_chan: &WebGLChan) -> bool {
```
nit: This probably doesn't need to be public at all.

MortimerGoro replied:
Yes, you are right. It was pub because I used it in the "packaging mode" I mentioned in the mailing list. But it can be private in the current implementation ;)

emilio commented on components/canvas/webgl_thread.rs:
```diff
@@ -0,0 +1,1191 @@
+                let (real_size, texture_id, _) = ctx.get_info();
+                self.observer.on_context_resize(context_id, texture_id, real_size);
+
+                let info = self.cached_context_info.get_mut(&context_id).unwrap();
```
nit: Can use `self.cached_context_info[&context_id]`, here and in a bunch of other places, to get nicer error messages.

MortimerGoro replied:
I dind't find a way to use the indexed accesor for mut references (Hashmap doesn't implement IndexMut and IndexSet is not ready yet AFAIK)

emilio replied:
Oh, ok.

emilio commented on components/canvas/webgl_thread.rs:
```diff
@@ -0,0 +1,1191 @@
+    fn handle_lock(&mut self, context_id: WebGLContextId, sender: WebGLSender<(u32, Size2D<i32>)>) {
+        let ctx = &self.contexts[&context_id];
+        if Some(context_id) != self.current_bound_webgl_context_id {
+            ctx.make_current();
```
Should we have a helper for this? We do have a few blocks like this. What about `make_current_if_needed(context_id)`.

emilio replied:
(It may return the current context, if the extra hashmap lookup worries you, I guess).

MortimerGoro replied:
Done

emilio commented on components/canvas/webgl_thread.rs:
```diff
@@ -0,0 +1,1191 @@
+
+/// Trait used by the generic WebGLExternalImageHandler implementation
+pub trait WebGLExternalImageApi {
+    fn lock(&mut self, ctx_id: WebGLContextId) -> (u32, Size2D<i32>);
```
Can you add docs about what does this do?

MortimerGoro replied:
Done

emilio reviewed (commented):
Looks reasonable to me, but I have a bunch of questions, and I've arguably only skimmed the more mechanical parts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #18025) made this pull request unmergeable. Please resolve the merge conflicts.

MortimerGoro commented:
¿should we have somebody else look at it? 

When review is ready ¿Do you want to squash commits or is good to keep the commit history in this case?

 Thanks!

glennw commented:
Thanks, those updates to my comments look good. I'm not sure what our policy is on squashing commits is - @jdm? Once @emilio is happy that his comments are addressed, r=me and let's get this merged!

jdm commented:
If the separate commits make sense (and build), I would prefer them to remain separate. Any commits that do not build should be squashed until they do.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17425) made this pull request unmergeable. Please resolve the merge conflicts.

MortimerGoro commented:
The separate commits may not build so I'll squash all commits after @emilio gives the ok

emilio commented:
If my comments and @glennw's are addressed, r=me,glennw

Thanks for all this @MortimerGoro!

@bors-servo delegate+

bors-servo commented:
:v: @MortimerGoro can now approve this pull request

MortimerGoro committed:
Improve WebGL architecture.

MortimerGoro commented:
@bors-servo r=glennw,emilio

bors-servo commented:
:pushpin: Commit 9d8485d has been approved by `glennw,emilio`

<!-- @bors-servo r=glennw,emilio 9d8485df06c5375a8eae21b7c5c249e8ab28b53e -->

bors-servo commented:
:hourglass: Testing commit 9d8485df06c5375a8eae21b7c5c249e8ab28b53e with merge 45bf52813b6e30dbe1279200cb544ce00ae278c5...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/8319)

MortimerGoro commented:
```
   Compiling constellation v0.0.1 (file:///home/servo/buildbot/slave/android/build/components/constellation)
fatal runtime error: allocator memory exhausted
error: Could not compile `script`.
```

larsbergstrom commented:
```
fatal runtime error: allocator memory exhausted
error: Could not compile `script`.
```

WOW :-)

MortimerGoro commented:
@bors-servo retry

jdm commented:
@bors-servo: retry

bors-servo commented:
:lock: Merge conflict

bors-servo commented:
:umbrella: The latest upstream changes (presumably #18090) made this pull request unmergeable. Please resolve the merge conflicts.

MortimerGoro commented:
Merge conflict fixed

@bors-servo r=glennw,emilio

bors-servo commented:
:pushpin: Commit fb7152f has been approved by `glennw,emilio`

<!-- @bors-servo r=glennw,emilio fb7152f74b6abdcab9cd36a2bc016117bb835a52 -->

bors-servo commented:
:hourglass: Testing commit fb7152f74b6abdcab9cd36a2bc016117bb835a52 with merge 7a603a6f0f58e38c71a1073152552db3bc66a631...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5467)

MortimerGoro commented:
Those tests failed because I accidentally removed a `byte_swap()` call in one of the review commits https://github.com/MortimerGoro/servo/commit/5cf6ac3243bce47460af2a67e560a331fd577a11

Let's try again!

@bors-servo r=glennw,emilio

MortimerGoro commented:
@bors-servo r=glennw,emilio

bors-servo commented:
:pushpin: Commit 703962f has been approved by `glennw,emilio`

<!-- @bors-servo r=glennw,emilio 703962fe61d673536eb982b45795ae13748f0f6a -->

bors-servo commented:
:hourglass: Testing commit 703962fe61d673536eb982b45795ae13748f0f6a with merge 90f55ea4580e2a15f7d70d0491444f18b972d450...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/8325), [arm32](http://build.servo.org/builders/arm32/builds/8550), [arm64](http://build.servo.org/builders/arm64/builds/8583), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8667), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5434), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5471), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8666), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1144), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1145), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5365), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5424), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1146), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1150), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3912)
Approved by: glennw,emilio
Pushing 90f55ea4580e2a15f7d70d0491444f18b972d450 to master...
